src/modules/flight_mode_manager/FlightModeManager.cpp

```diff
@@ -215,6 +215,29 @@ void FlightModeManager::start_flight_task()
 			_task_failure_count = 0;
 		}
 
+		// Take-over landing from navigator if precision landing is enabled
+
+	} else if ((_vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_RTL ||
+		    _vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LAND)
+		   && _param_rtl_pld_md.get() > 0) {
```

Comment:
Does using this flag for both RTL and Land have any unintended consequences? What if you want required precision land for RTL but not for land? I can definitely see a case where you might have a landing target back home you are required to land on, but during a flight you may want to trigger a land for safety/other reasons. In this case the land logic would be broken.

I see Dusan already pointed this out but not all AUTO_LANDs will want to be precision lands so we need to somehow pass the state of `land_precision` [in navigator ](https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/mavlink/mavlink_mission.cpp#L1384-L1389)to this flight task to handle Mission NAV_CMD_LAND precision lands.

Reply:
hehe just found this https://docs.google.com/document/d/1N59JHFqL_E9X4b-U0CzxW_eydqXUHZMxAH_FTD30ilk/edit

>Refactor Navigator into having one FlightMode per navigation state

I think this implies any behavior that is shared between modes and mission items need to be handled as a FlightTask (eg RTL, Land, Precision Land, Takeoff, Loiter, etc). Am I understanding this correctly? Maybe we can publish a new uORB topic `navigation_mode` from navigator which the `FlightModeManager` listens to and switches to the correct task by checking

For mission items
```
_vehicle_control_mode_sub.get().flag_control_auto_enabled && (_nav_mode_sub.get().navigation_mode == MODE_OF_INTEREST)
```

For modes
```
_vehicle_status_sub.get().nav_state == vehicle_status_s::NAVIGATION_STATE_*
```

This will kill off nearly all of the MissionBlock classes in navigator and consolidate control code to a single place (FlightTasks)

Reply:
> What if you want required precision land for RTL but not for land
That feature does not exist yet :sweat_smile: 

Assuming I manage to make it work with missions, you would then have to set `RTL_PLD_MD=0` to disable precision landing, and then specifically enable it for a mission item. That's the only way I can think of that does not involve yet another parameter for precision LAND besides precision RTL.

Reply:
> I think allowing precision land for the land command (NAVIGATION_STATE_AUTO_LAND) is a good idea but it should be a separate parameter to enable it.

I'm just worried that we end up creating too many parameters for something simple. 

There's also the DESCEND flight mode, which does not control position, but only velocity. Usually used when GPS is lost mid-flight. Theoretically DESCEND could still perform precision landing as long as the target is visible, even without GPS.

So I think what all of this tells me is that I'll now make all existing precland use cases work (precland as part of RTL, flight mode or mission) and then in a second iteration actually improve the architecture as you suggested.

Reply:
@dakejahl 
> What if you want required precision land for RTL but not for land

I thought about it, and in my opinion we should get rid of the `RTL_PLD_MD` parameter in the first place. Otherwise we would have to also add a parameter for precision land mode `LND_PLD_MD` or whatever, and perhaps for other modes as well where you wish to have the precision-target being used somehow.

Down the line I would just make them separate flight modes. Then we would have
- RTL
- PrecisionRTL
- Land
- PrecisionLand
- whatever else you can think of

This would also reduce the number of code paths for how one can end up in PrecisionLand, which right now is very confusing.

Reply:
makes sense, also makes it more clear about what a mode is and how you get into it